asserts/tool/sign.go

```diff
@@ -0,0 +1,249 @@
+}
+
+// Sign produces the text of a signed assertion as specified by req.
+func Sign(req *SignRequest, keypairMgr asserts.KeypairManager) ([]byte, error) {
+	typ := asserts.Type(req.AssertionType)
+	if typ == nil {
+		return nil, fmt.Errorf("invalid assertion type: %q", req.AssertionType)
+	}
+	if req.Revision < 0 {
+		return nil, fmt.Errorf("assertion revision cannot be negative")
+	}
+	if req.AccountKey == nil && req.KeyID == "" {
+		return nil, fmt.Errorf("both account-key and key id were not specified")
+	}
+
+	var nestedStatement nestedStatement
+	err := parseStatement(req, &nestedStatement)
+	if err != nil {
+		return nil, err
+	}
+	if nestedStatement.Headers == nil {
+		// flat headers, reparse
+		err := parseStatement(req, &nestedStatement.Headers)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	headers, err := stringify(nestedStatement.Headers)
+	if err != nil {
+		return nil, err
+	}
+	body := []byte(nestedStatement.Body)
+
+	keyID := req.KeyID
+	authorityID := req.AuthorityID
+
+	if req.AccountKey != nil {
+		if keyID != "" || authorityID != "" {
+			return nil, fmt.Errorf("cannot mix specifying an account-key together with key id and/or authority-id")
+		}
+
+		// use the account-key as a handle to get the information about
+		// signer and key id
+		a, err := asserts.Decode(req.AccountKey)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse handle account-key: %v", err)
+		}
+		accKey, ok := a.(*asserts.AccountKey)
+		if !ok {
+			return nil, fmt.Errorf("cannot use handle account-key, not actually an account-key, got: %s", a.Type().Name)
+		}
+		keyID = accKey.PublicKeyID()
+		authorityID = accKey.AccountID()
+
+		// extra sanity checks about fingerprint
+		pk, err := keypairMgr.Get(authorityID, keyID)
+		if err != nil {
+			return nil, err
+		}
+		expFpr := accKey.PublicKeyFingerprint()
+		gotFpr := pk.PublicKey().Fingerprint()
+		if gotFpr != expFpr {
+			return nil, fmt.Errorf("cannot use found private key, fingerprint does not match account-key, expected %q, got: %s", expFpr, gotFpr)
+		}
+	}
+
+	if authorityID != "" {
+		headers["authority-id"] = authorityID
+	}
+
+	if headers["authority-id"] == "" {
+		return nil, fmt.Errorf("cannot sign assertion with unspecified signer identifier (aka authority-id)")
+	}
+
+	if req.Revision != 0 {
+		headers["revision"] = strconv.Itoa(req.Revision)
+	}
+
+	adb, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		KeypairManager: keypairMgr,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	a, err := adb.Sign(typ, headers, body, keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	return asserts.Encode(a), nil
+}
+
+// stringify lets the invoker use a limited amount of structured input
+// without having to convert everything obvious to strings upfront on
+// their side, we convert integers, bool (to yes|no), list of strings
+// (to comma separated) and nil (to empty).
+func stringify(m map[string]interface{}) (map[string]string, error) {
```

Comment:
Very concerned about this one. Sounds completely obvious and innocent, yet it's implicitly defining critical properties about the format of headers. Back then we've constrained the headers to map[string]string precisely so we could not worry by consciously keeping it to a minimal set of types until a further time. This one function is now opening up to nested structures with no discussion about how structured documents should be put in place.

My suggestion is this: to avoid the discussion today, let's keep the tool as taking a map[string]string type. Then, on a follow up let's open the header format itself to be map[string]interface{}, and let's design how structured documents look like both in memory and when serialized.